# Introduction to Pandas

## Pandas provides Python data frames

* Popular and established
* Inspired by R dataframes
* Built on `numpy` for fast computation

In [1]:
import pandas as pd

## Our first dataframe

In [2]:
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "Love_of_R": [2, 5, 11],
                   "years_at_wsu": [4, 17, 5]})
df.head()

Names  Python_mastery  Love_of_R  years_at_wsu
0  Iverson            10.0          2             4
1   Malone             5.0          5            17
2   Bergen             1.0         11             5

## Reading from a csv

* Most data sets will be read in from a csv or JSON data file
* `Pandas` provides `read_csv` and `read_json`

### Open a local file w/ relative path

In [42]:
# Won't work in colab
artist = pd.read_csv('./data/Artists.csv')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

### Open a web address

In [3]:
url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.csv"
artists =  pd.read_csv(url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

# JSON data file

* Another (more modern) storage
* Here the data is stored in row `dict`

```{json}
[
{
  "ConstituentID": 1,
  "DisplayName": "Robert Arneson",
  "ArtistBio": "American, 1930–1992",
  "Nationality": "American",
  "Gender": "Male",
  "BeginDate": 1930,
  "EndDate": 1992,
  "Wiki QID": null,
  "ULAN": null
},
{
  "ConstituentID": 2,
  "DisplayName": "Doroteo Arnaiz",
  "ArtistBio": "Spanish, born 1936",
  "Nationality": "Spanish",
  "Gender": "Male",
  "BeginDate": 1936,
  "EndDate": 0,
  "Wiki QID": null,
  "ULAN": null
},
...
```

## `pandas` can read `json` data

In [4]:
# Won't work in colab
artists =  pd.read_json('./data/Artists.json')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

In [5]:
json_url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.json"
artists =  pd.read_json(json_url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

## <font color="red"> Exercise 2.1.2 </font>
    
Use tab-completion and `help` to discover and explore two more methods of reading a file into a `Pandas` dataframe.


In [14]:
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers:

read_csv(filepath_or_buffer: Union[str, pathlib.Path, IO[~AnyStr]], sep=',', delimiter=None, header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None, compression='infer', thousands=None, decimal: str = '.', lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None, comment=None, encoding=None, dialect=None, error_bad_lines=True, warn_bad_lines=True, delim_whitespace=False, low_memory=True, memory_map=False, float_precision=None)
    Read a comma-separated values (csv) file in

In [9]:
#can read in .csv files

artists_csv = pd.read_csv('./data/Artists.csv')

artists_csv.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

In [15]:
help(pd.read_table)

Help on function read_table in module pandas.io.parsers:

read_table(filepath_or_buffer: Union[str, pathlib.Path, IO[~AnyStr]], sep='\t', delimiter=None, header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None, compression='infer', thousands=None, decimal: str = '.', lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None, comment=None, encoding=None, dialect=None, error_bad_lines=True, warn_bad_lines=True, delim_whitespace=False, low_memory=True, memory_map=False, float_precision=None)
    Read general delimited file into DataF

In [13]:
#can specify file format manually

artists_freestly = pd.read_table('./data/Artists.csv', delimiter=",")

artists_freestly.head()

pd.read

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

> Discuss what you found here

Pandas appears to be setup to read most standard file types by defualt, doing everything from sas, html, and even pickle. Most commands for these are fairly startiforward and follow similar patterns. Additionally, for the cases where the format is more niche, there is the read_table option where you can specify most options yourself. Above, I demonstrated that it can replicate the .csv variant easily enough.

## <font color="red"> Exercise 2.1.2 </font>
    
Read in the `Artwork.csv` from [https://github.com/MuseumofModernArt/collection](https://github.com/MuseumofModernArt/collection) and display the head of the resulting dataframe.


In [20]:
# Your code here

artwork_git = pd.read_csv('https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artworks.csv')

artwork_git.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

## So what is a `DateFrame`

* Like R, Pandas focuses on columns
* Think `dict` of `(str, Series)` pairs 
* A series is a typed list-like structure

In [62]:
# This is how I imagine a dataframe
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "years_at_wsu": [4.5, 17.5, 5.5]})

## Columns are `Series` and hold one type of data

In [63]:
type(artists.BeginDate), type(artists.DisplayName)

(pandas.core.series.Series, pandas.core.series.Series)

In [64]:
artists.BeginDate.dtype, artists.DisplayName.dtype

(dtype('int64'), dtype('O'))

## Two ways to access a column

* **Method 1:** like a dictionary
    * `df["column_name"]`
* **Method 2:** like an object attribute
    * `df.column_name`
    * Only for proper names!

In [65]:
artists.BeginDate.head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

In [66]:
artists['BeginDate'].head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

## More on data types

* See all data types with `df.dtypes`
* You can set the `dtypes` when you read a dataframe
* Read more about types: [Pandas docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dtypes.html)

In [52]:
artists.dtypes

ConstituentID      int64
DisplayName       object
ArtistBio         object
Nationality       object
Gender            object
BeginDate          int64
EndDate            int64
Wiki QID          object
ULAN             float64
dtype: object

## Setting `dtypes` with `read_csv`

We can pass a `dict` of types to `dtype` keyword

In [56]:
import numpy as np
artist_types = {'ConstituentID': np.int64,
                'DisplayName': str,
                'ArtistBio': str,
                'Nationality': str,
                'Gender':str,
                'BeginDate': np.int64,
                'EndDate': np.int64,
                'Wiki QID': str,
                'ULAN':pd.Int64Dtype()} # If you get an error ==> update pandas (see below)
artists2 = pd.read_csv('./data/artists.csv', dtype = artist_types)
artists2.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID       ULAN  
0       1930     1992       NaN       <NA>  
1       1936        0       NaN       <NA>  
2       1941        0       NaN       <NA>  
3       1946        0  Q1063584  500027998  
4       1941        0       NaN       <NA>

## What's up with `ULAN` ?

* Currently, `numpy` $\rightarrow$ no missing `int`s
* Pandas correct this with the `pd.Int64Dtype()` type
    * Only available in `pandas >= 0.24.0`

In [57]:
pd.__version__

'1.1.3'

## An `Int` by any other name ...

* `np.int64` $\rightarrow$ no missing values
* `pd.Int64Dtype()` $\rightarrow$ allows `NaN`

In [58]:
artists2.dtypes

ConstituentID     int64
DisplayName      object
ArtistBio        object
Nationality      object
Gender           object
BeginDate         int64
EndDate           int64
Wiki QID         object
ULAN              Int64
dtype: object

## Preview of coming attractions

* Now we can switch `BeginDate` and `EndDate` from `0` to `np.NaN`
* We will do this in the next section

# Getting to know your data

## Basic inspection tools

* `df.head()`        first five rows
* `df.tail()`        last five rows
* `df.sample(5)`     random sample of rows
* `df.shape`         number of rows/columns in a tuple
* `df.describe()`    calculates measures of central tendency
* `df.info()`

## <font color="red"> Exercise 1: Inspect the artwork from MoMA </font>

#### Read the csv and inspect the `head`

In [67]:
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

**Task:** Write a few sentences describing an problems

The first thing that immediatley jumps out at me is the large amount of missing data, which is likely going to break fair amount of functions on its own and is also presneted inconsistantly. Somtimes its blank, sometimes its "()", and sometimes its NaN.

Secondondly, while its is nice that medium is specific, it really ought yo be recoded or possibly even split into new columns to be both more consistant and actually useful.

Lastly, and probably the worst thing by far, is the agregious breaking of the golden in rule in several columns, notably the title and bio, which not only present information that should be on its own, but also does so inconisitantly.

#### Inspect the column names with the `columns` attribute

In [17]:
artwork.columns

Index(['Title', 'Artist', 'ConstituentID', 'ArtistBio', 'Nationality',
       'BeginDate', 'EndDate', 'Gender', 'Date', 'Medium', 'Dimensions',
       'CreditLine', 'AccessionNumber', 'Classification', 'Department',
       'DateAcquired', 'Cataloged', 'ObjectID', 'URL', 'ThumbnailURL',
       'Circumference (cm)', 'Depth (cm)', 'Diameter (cm)', 'Height (cm)',
       'Length (cm)', 'Weight (kg)', 'Width (cm)', 'Seat Height (cm)',
       'Duration (sec.)'],
      dtype='object')

**Question:** See any problems?

*Your thoughts here*

Dimensions seems redudant given that we already have columns breaking that apart futher on, the names having spaces is simply annoying, and some of the names (such as Duration) could probably be made more explicit. As discussed above, there should likely be a few more columns here then there presently is.

#### Inspect the tail

In [11]:
artwork.tail()

Title           Artist  \
136526   Duplicate of plate facing page 6 from Mazas  Maximilien Luce   
136527   Duplicate of plate facing page 7 from Mazas  Maximilien Luce   
136528   Duplicate of plate facing page 8 from Mazas  Maximilien Luce   
136529   Duplicate of plate facing page 9 from Mazas  Maximilien Luce   
136530  Duplicate of plate facing page 10 from Mazas  Maximilien Luce   

       ConstituentID            ArtistBio Nationality BeginDate EndDate  \
136526          3621  (French, 1858–1941)    (French)    (1858)  (1941)   
136527          3621  (French, 1858–1941)    (French)    (1858)  (1941)   
136528          3621  (French, 1858–1941)    (French)    (1858)  (1941)   
136529          3621  (French, 1858–1941)    (French)    (1858)  (1941)   
136530          3621  (French, 1858–1941)    (French)    (1858)  (1941)   

        Gender  Date                                             Medium  \
136526  (Male)  1894  Lithograph from the supplementary suite of an ...   
136527  (Male)  1894  Lithograph from the supplementary suite of an ...   
136528  (Male)  1894  Lithograph from the supplementary suite of an ...   
136529  (Male)  1894  Lithograph from the supplementary suite of an ...   
136530  (Male)  1894  Lithograph from the supplementary suite of an ...   

             ...       ThumbnailURL Circumference (cm) Depth (cm)  \
136526       ...                NaN                NaN        NaN   
136527       ...                NaN                NaN        NaN   
136528       ...                NaN                NaN        NaN   
136529       ...                NaN                NaN        NaN   
136530       ...                NaN                NaN        NaN   

       Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
136526           NaN        32.3         NaN         NaN        23.5   
136527           NaN        30.4         NaN         NaN        24.0   
136528           NaN        33.0         NaN         NaN        24.2   
136529           NaN        32.0         NaN         NaN        24.0   
136530           NaN        22.0         NaN         NaN        30.6   

       Seat Height (cm) Duration (sec.)  
136526              NaN             NaN  
136527              NaN             NaN  
136528              NaN             NaN  
136529              NaN             NaN  
136530              NaN             NaN  

[5 rows x 29 columns]

#### Check out the `shape`

In [15]:
artwork.shape

(136531, 29)

**Question:** What do these number mean?

*Your thoughts here*

Therre are 136,531 rows amd 29 columns in the data frame.

#### Use `describe` to compute statistics

In [13]:
artwork.describe()

ObjectID  Circumference (cm)    Depth (cm)  Diameter (cm)  \
count  136531.000000           10.000000  13157.000000    1429.000000   
mean    90576.707253           44.868020     16.789801      23.184666   
std     68230.367311           28.631604     55.337322      45.070383   
min         2.000000            9.900000      0.000000       0.635000   
25%     36117.500000           23.500000      0.000000       7.900000   
50%     72981.000000           36.000000      0.500000      13.800000   
75%    137842.500000           71.125000     10.200000      24.900000   
max    294767.000000           83.800000   1808.483617     914.400000   

         Height (cm)  Length (cm)    Weight (kg)     Width (cm)  \
count  117187.000000   738.000000     289.000000  116280.000000   
mean       37.642548    89.892356    1287.944097      38.072156   
std        47.700061   330.290367   12038.129595      66.482430   
min         0.000000     0.000000       0.090000       0.000000   
25%        18.100000    17.100000       5.670000      17.780036   
50%        27.940056    26.700000      19.958300      25.400051   
75%        44.300000    79.700000      80.286700      44.600000   
max      9140.000000  8321.056600  185067.585957    9144.000000   

       Seat Height (cm)  Duration (sec.)  
count               0.0     3.304000e+03  
mean                NaN     7.766900e+03  
std                 NaN     1.145070e+05  
min                 NaN     0.000000e+00  
25%                 NaN     2.380000e+02  
50%                 NaN     7.800000e+02  
75%                 NaN     4.320000e+03  
max                 NaN     6.283065e+06

#### Use `info` to look at types and totals

In [17]:
artwork.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136531 entries, 0 to 136530
Data columns (total 29 columns):
Title                 136492 non-null object
Artist                135083 non-null object
ConstituentID         135083 non-null object
ArtistBio             131085 non-null object
Nationality           135083 non-null object
BeginDate             135083 non-null object
EndDate               135083 non-null object
Gender                135083 non-null object
Date                  134128 non-null object
Medium                125364 non-null object
Dimensions            125536 non-null object
CreditLine            133653 non-null object
AccessionNumber       136531 non-null object
Classification        136531 non-null object
Department            136531 non-null object
DateAcquired          129775 non-null object
Cataloged             136531 non-null object
ObjectID              136531 non-null int64
URL                   78575 non-null object
ThumbnailURL          67825 non-null

**Question:** What did you learn from the last two cells?

*Your thoughts here*

Beyond some interesting, if basic summations of the data for rows it could be computed for, we also learn the datatypes that pandas currently identifies the columns as. In the case of this particular dataframe, most are objects (strings) with a few float64s thrown in for the dimension variables.

There are also a few oddities that could use some explanation, leaving us with questions such as "is there a reason most dimensions are barely filled out" and "why does seat Height not contain any data at all." I works well to provide a snapshot of the defficiencies found within, but also of the possible knowledge one could learn if the data is placed in a more desireable form ripe for analysis.

The main takeaway from all of this is that we should probably spend some time with this data set in openrefine and possibly contact the curators before we begin anything serious with this chunk of data.